Add Jinja2 support

heyman opened the pull request:
Hi!

This PR adds a jinja2 extension that adds the `{% importmap_scripts %}`  tag to jinja2 templates.

It also adds a section about Jinja2 support to the Readme, and it adds two additional tests that uses jinja.

heyman committed:
Add Jinja2 extension with the {% importmap_scripts %}  tag

davegaeddert commented on importmap/jinja2/extension.py:
```diff
@@ -0,0 +1,32 @@
+from importmap.templatetags.importmap import importmap_scripts
+
+
+SCRIPT_TEMPLATE = """
```
Do you know how to put this in a template file (that can be overridden by a user)? Would it be `templates/jinja2/importmap/scripts.html`? https://docs.djangoproject.com/en/4.1/topics/templates/#module-django.template.loader

May be overkill, but I'd like it in a place that can be overridden for now because I haven't done anything with settings around the shim (version, whether or not to include it).

heyman replied:
Ah, I was wondering why a template file was used for the scripts. Didn't think about the fact that it allows users to override it, but it makes sense.

However, from the top of my head I can't think of a simple way to achieve the same thing with Jinja. (since it depends on how the user has configured jinja in their `TEMPLATES` setting key).

heyman replied:
One possible solution that I can think of would be to use *Django's* templating system to render the scripts-snippet. Changing the `_render` function to something like:

```python
from django.template import Engine, Context

    ...
    def _render(self, _):
        template = Engine.get_default().get_template("importmap/scripts.html")
        return template.render(Context(importmap_scripts()))
```

That should allow users to override the file, but maybe a bit strange that one would use a Django template for that when using the Jinja extension?

davegaeddert reviewed (changes requested):
Thanks @heyman! I use Jinja a lot, but not with Django — seems like a nice addition. I've got one small request but other than that this looks pretty good and simple.

heyman committed:
Allow users to customize <script> code that gets included by {% importmap_scripts %} when using the Jinja2 extension.

This uses Django's template engine to render the same importmap/scripts.html template.

heyman commented:
Okay, I've now changed so that the Jinja2 extension uses Django's template engine to render the `importmap/scripts.html` which should allow end-users to override it. I think everyone who uses Jinja2 templates with Django will still have a Django template engine in their config (needed for the Admin, and every other app that comes with templates), so I think it shouldn't be a problem (other than that people might find it a bit unintuitive, but since this is a non-default and undocumented feature, I think that's okay).

davegaeddert committed:
Move jinja to an environment example

davegaeddert committed:
Merge branch 'jinja' of https://github.com/heyman/django-importmap into jinja

* 'jinja' of https://github.com/heyman/django-importmap:
  Allow users to customize <script> code that gets included by {% importmap_scripts %} when using the Jinja2 extension.

davegaeddert commented:
Ah, sorry I sort of pulled a 180 while you were working on this apparently. I'm not actually sure I want to support/maintain a Jinja extension in here directly... after reading the Django docs a little more, it seems like the recommended approach for stuff like this is to customize the environment? https://docs.djangoproject.com/en/4.1/topics/templates/#django.template.backends.jinja2.Jinja2

I agree the loading a Django template inside Jinja feels kinda weird, even though it sounds like Django templates should generally be available still.

How do you feel about the Jinja process being more of a "guide" than anything (do you personally already have a custom environment set)? I made it a little easier to use via `Importmap.json(development=<True|False>)`: https://github.com/dropseed/django-importmap/blob/a4e169205ebff112f231e5456bf7e3f50c835fb3/README.md#using-jinja2

davegaeddert committed:
Merge branch 'master' into jinja

* master:
  Separate load importmap from generate

heyman commented:
To me it feels a bit easier (as an end-user) to just add an additional Jinja2 extension and then you're all set (and with the same API as for the Django template engine). But injecting just the JSON blob into globals works as well I guess.

> do you personally already have a custom environment set

Yes, in every Django project where I use Jinja2 I've set up a custom environment.

davegaeddert commented:
Ok, I'm going to say that, for now, putting it in the Jinja environment is going to be the suggestion. I just don't expect there to be much Jinja usage, so I don't want supporting an extension to make things more complicated for whatever reason. Definitely happy to revisit it though if more Jinja people want to use it and complain!

heyman commented:
Ok!

Actually, while thinking about a different issue (#16) I realized that having access to the importmap JSON data might actually be better for my current use-case.
